File: backend/routers/courses.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from fastapi import APIRouter, HTTPException
from database import get_conn, enrich

router = APIRouter(prefix="/api/courses", tags=["courses"])

VALID_COURSES = {"btech", "mba", "polytechnic", "arts"}

COURSE_META = [
    {"id": "btech",       "name": "B.Tech",     "description": "Engineering programs in various specializations"},
    {"id": "mba",         "name": "MBA",         "description": "Management and business administration"},
    {"id": "polytechnic", "name": "Polytechnic", "description": "Diploma courses in engineering"},
    {"id": "arts",        "name": "BA/BSc/BCom", "description": "Arts, Science, and Commerce programs"},
]


@router.get("/")
def list_courses():
    return COURSE_META


@router.get("/{course_type}/colleges")
def colleges_by_course(course_type: str):
    if course_type not in VALID_COURSES:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid course. Valid options: {', '.join(VALID_COURSES)}"
        )
    conn = get_conn()
    rows = conn.execute("""
        SELECT c.* FROM colleges c
        INNER JOIN college_courses cc ON c.id = cc.college_id
        WHERE cc.course = ?
    """, (course_type,)).fetchall()
    result = [enrich(conn, dict(r)) for r in rows]
    conn.close()
    return result

File: backend/routes/courses.js
const express = require('express');
const router = express.Router();
const db = require('../db');

const VALID_COURSES = ['btech', 'mba', 'polytechnic', 'arts'];

// GET /api/courses - list all available course types
router.get('/', (req, res) => {
  const courses = [
    { id: 'btech', name: 'B.Tech', description: 'Engineering programs in various specializations' },
    { id: 'mba', name: 'MBA', description: 'Management and business administration' },
    { id: 'polytechnic', name: 'Polytechnic', description: 'Diploma courses in engineering' },
    { id: 'arts', name: 'BA/BSc/BCom', description: 'Arts, Science, and Commerce programs' }
  ];
  res.json(courses);
});

// GET /api/courses/:courseType/colleges - colleges offering a specific course
router.get('/:courseType/colleges', (req, res) => {
  const { courseType } = req.params;

  if (!VALID_COURSES.includes(courseType)) {
    return res.status(400).json({ error: `Invalid course type. Valid types: ${VALID_COURSES.join(', ')}` });
  }

  const colleges = db.prepare(`
    SELECT c.* FROM colleges c
    INNER JOIN college_courses cc ON c.id = cc.college_id
    WHERE cc.course = ?
  `).all(courseType);

  const enriched = colleges.map(college => {
    college.courses = db.prepare('SELECT course FROM college_courses WHERE college_id = ?')
      .all(college.id).map(r => r.course);
    college.branches = db.prepare('SELECT branch FROM college_branches WHERE college_id = ?')
      .all(college.id).map(r => r.branch);
    college.facilities = db.prepare('SELECT facility FROM college_facilities WHERE college_id = ?')
      .all(college.id).map(r => r.facility);
    return college;
  });

  res.json(enriched);
});

module.exports = router;
